File: src_cpp/ai_factory/cpu/cir_plus_plus/bermudan_swaptions.cpp
#include "ai_factory/cpu/cir_plus_plus/bermudan_swaptions.hpp"

#include "ai_factory/common/fixed_income/cir_plus_plus.hpp"
#include "ai_factory/cpu/common/bermudan_lsm.hpp"
#include "ai_factory/cpu/common/philox.hpp"

#include <algorithm>
#include <array>
#include <cmath>
#include <stdexcept>
#include <vector>

namespace ai_factory::cpu::cir_plus_plus {
namespace {

constexpr int kMaxExercises = 8;
constexpr int kMaxPayments = 20;
constexpr double kBasisRateScale = 0.04;
constexpr double kQePsiCutoff = 1.5;

void validate(const cuda::BermudanSwaptionTerms& product) {
    if (product.exercise_count < 2 || product.exercise_count > kMaxExercises
        || product.payment_count < product.exercise_count + 2
        || product.payment_count > kMaxPayments
        || std::abs(product.exercise_period - product.accrual_period) > 1.0e-12) {
        throw std::invalid_argument("Invalid Bermudan swaption schedule.");
    }
}

}  // namespace

void price_bermudan_swaption_batch(
    const cuda::CirPlusPlusBermudanSwaptionRow* rows,
    std::size_t row_count,
    std::size_t num_paths,
    double target_dt,
    cuda::MonteCarloOutput* outputs
) {
    if (num_paths < 2U || target_dt <= 0.0) {
        throw std::invalid_argument("Invalid CIR Bermudan Monte Carlo settings.");
    }
#ifdef _OPENMP
#pragma omp parallel for schedule(static)
#endif
    for (std::ptrdiff_t row_index = 0;
         row_index < static_cast<std::ptrdiff_t>(row_count);
         ++row_index) {
        const auto& row = rows[row_index];
        const auto& model = row.model;
        const auto& product = row.product;
        validate(product);
        const auto exercise_count = static_cast<std::size_t>(product.exercise_count);
        const double last_exercise = product.first_exercise
            + (product.exercise_count - 1) * product.exercise_period;
        const auto num_steps = static_cast<std::size_t>(std::max(
            1LL, std::llround(last_exercise / target_dt)
        ));
        const double dt = last_exercise / static_cast<double>(num_steps);

        std::array<std::size_t, kMaxExercises> exercise_steps{};
        std::array<std::array<double, kMaxPayments>, kMaxExercises> bond_a{};
        std::array<std::array<double, kMaxPayments>, kMaxExercises> bond_b{};
        for (std::size_t exercise = 0; exercise < exercise_count; ++exercise) {
            const double exercise_time = product.first_exercise
                + exercise * product.exercise_period;
            exercise_steps[exercise] = static_cast<std::size_t>(
                std::llround(exercise_time / dt)
            );
            for (int payment = static_cast<int>(exercise);
                 payment < product.payment_count;
                 ++payment) {
                const double maturity = product.first_exercise
                    + (payment + 1) * product.accrual_period;
                bond_a[exercise][payment] = fixed_income::cir_plus_plus_bond_a(
                    exercise_time, maturity, model.initial_factor,
                    model.kappa, model.theta, model.volatility,
                    row.curve.beta0, row.curve.beta1,
                    row.curve.beta2, row.curve.tau
                );
                bond_b[exercise][payment] = fixed_income::cir_bond_b(
                    model.kappa, model.volatility, maturity - exercise_time
                );
            }
        }

        const double decay = std::exp(-model.kappa * dt);
        const double one_minus_decay = 1.0 - decay;
        const double volatility_squared = model.volatility * model.volatility;
        std::vector<double> states(exercise_count * num_paths);
        std::vector<double> discounts(exercise_count * num_paths);
        for (std::size_t path = 0; path < num_paths; ++path) {
            simulation::PhiloxNormalSequence normals(row.seed, path);
            simulation::PhiloxUniformSequence uniforms(row.seed, path + num_paths);
            double rate = model.initial_factor;
            double integral = 0.0;
            std::size_t next_exercise = 0U;
            for (std::size_t step = 1U; step <= num_steps; ++step) {
                const double previous_rate = rate;
                const double mean = model.theta + (rate - model.theta) * decay;
                const double variance =
                    rate * volatility_squared * decay * one_minus_decay / model.kappa
                    + model.theta * volatility_squared * one_minus_decay
                          * one_minus_decay / (2.0 * model.kappa);
                const double psi = variance / (mean * mean);
                const double normal = normals.next();
                const double uniform = uniforms.next();
                if (psi <= kQePsiCutoff) {
                    const double inverse_psi = 1.0 / psi;
                    const double b_squared = 2.0 * inverse_psi - 1.0
                        + std::sqrt(2.0 * inverse_psi)
                              * std::sqrt(std::max(2.0 * inverse_psi - 1.0, 0.0));
                    const double shifted = std::sqrt(b_squared) + normal;
                    rate = mean / (1.0 + b_squared) * shifted * shifted;
                } else {
                    const double probability = (psi - 1.0) / (psi + 1.0);
                    rate = uniform <= probability
                        ? 0.0
                        : -mean * std::log((1.0 - uniform) / (1.0 - probability))
                              / (1.0 - probability);
                }
                integral += 0.5 * (previous_rate + rate) * dt;
                if (next_exercise < exercise_count
                    && step == exercise_steps[next_exercise]) {
                    const std::size_t index = next_exercise * num_paths + path;
                    states[index] = rate;
                    const double exercise_time = product.first_exercise
                        + next_exercise * product.exercise_period;
                    discounts[index] = fixed_income::cir_plus_plus_path_discount(
                        integral, exercise_time, model.initial_factor,
                        model.kappa, model.theta, model.volatility,
                        row.curve.beta0, row.curve.beta1,
                        row.curve.beta2, row.curve.tau
                    );
                    ++next_exercise;
                }
            }
        }

        const auto evaluate = [&](std::size_t exercise, double rate) {
            double annuity = 0.0;
            for (int payment = static_cast<int>(exercise);
                 payment < product.payment_count;
                 ++payment) {
                annuity += product.accrual_period * bond_a[exercise][payment]
                    * std::exp(-bond_b[exercise][payment] * rate);
            }
            const int last_payment = product.payment_count - 1;
            const double end_bond = bond_a[exercise][last_payment]
                * std::exp(-bond_b[exercise][last_payment] * rate);
            const double signed_swap = static_cast<double>(product.direction)
                * (1.0 - end_bond - product.fixed_rate * annuity);
            return lsm::ExerciseValue{
                product.notional * std::max(signed_swap, 0.0),
                ((1.0 - end_bond) / annuity) / kBasisRateScale,
            };
        };
        const auto cashflows = lsm::bermudan_cashflows_present_value(
            states, discounts, exercise_count, num_paths, evaluate
        );
        double sum = 0.0;
        double sumsq = 0.0;
        for (const double cashflow : cashflows) {
            sum += cashflow;
            sumsq += cashflow * cashflow;
        }
        const double n = static_cast<double>(num_paths);
        const double mean = sum / n;
        const double variance = (sumsq - n * mean * mean) / (n - 1.0);
        outputs[row_index] = {mean, std::sqrt(std::max(variance, 0.0) / n)};
    }
}

}  // namespace ai_factory::cpu::cir_plus_plus

File: src_python/ai_factory/pytorch/cir/bermudan_swaptions.py
"""Batched CIR QE Bermudan swaption Monte Carlo."""

from __future__ import annotations

from time import perf_counter

import torch

from ai_factory.pytorch.common.bermudan_lsm import present_value_cashflows
from ai_factory.pytorch.common.device import resolve_device, synchronize, warmup
from ai_factory.pytorch.common.fixed_income import cir_bond_coefficients
from ai_factory.pytorch.common.monte_carlo import price_summary

DEFAULT_CPU_BATCH_ROWS = 16
DEFAULT_CUDA_BATCH_ROWS = 64
MAX_EXERCISES = 8
MAX_PAYMENTS = 20
QE_PSI_CUTOFF = 1.5
BASIS_RATE_SCALE = 0.04


def _column(values, target):
    return torch.tensor(values, device=target, dtype=torch.float64)[:, None]


def _price_batch_impl(rows, models, products, *, num_paths, target, target_dt, batch_rows):
    started = perf_counter()
    outputs = []
    payment_axis = torch.arange(MAX_PAYMENTS, device=target)[None, :]
    exercise_axis = torch.arange(MAX_EXERCISES, device=target)[None, :]
    for offset in range(0, len(rows), batch_rows):
        batch = rows[offset : offset + batch_rows]
        ms = [models[row["model_id"]] for row in batch]
        ps = [products[row["product_id"]] for row in batch]
        size = len(batch)
        kappa = _column([m["kappa"] for m in ms], target)
        theta = _column([m["theta"] for m in ms], target)
        sigma = _column([m["volatility"] for m in ms], target)
        rate = _column([m["initial_rate"] for m in ms], target).expand(-1, num_paths).clone()
        first = _column([p["first_exercise"] for p in ps], target)
        period = _column([p["exercise_period"] for p in ps], target)
        accrual = _column([p["accrual_period"] for p in ps], target)
        strike = _column([p["fixed_rate"] for p in ps], target)
        notional = _column([p["notional"] for p in ps], target)
        direction = _column([p["direction"] for p in ps], target)
        exercise_count = torch.tensor([p["exercise_count"] for p in ps], device=target)
        payment_count = torch.tensor([p["payment_count"] for p in ps], device=target)
        exercise_times = first + period * exercise_axis
        schedule: dict[int, list[tuple[int, int]]] = {}
        for row_index, product in enumerate(ps):
            for exercise in range(product["exercise_count"]):
                step = round(
                    (product["first_exercise"] + exercise * product["exercise_period"])
                    / target_dt
                )
                schedule.setdefault(step, []).append((row_index, exercise))
        max_step = max(schedule)
        dt = target_dt
        decay = torch.exp(-kappa * dt)
        one_minus_decay = 1.0 - decay
        sigma_squared = sigma.square()
        integral = torch.zeros_like(rate)
        states = torch.zeros((size, num_paths, MAX_EXERCISES), device=target, dtype=torch.float64)
        discounts = torch.ones_like(states)
        for step in range(1, max_step + 1):
            previous = rate
            mean = theta + (rate - theta) * decay
            variance = (
                rate * sigma_squared * decay * one_minus_decay / kappa
                + theta * sigma_squared * one_minus_decay.square() / (2.0 * kappa)
            )
            psi = variance / mean.square()
            normal = torch.randn_like(rate)
            uniform = torch.rand_like(rate)
            inverse_psi = 1.0 / torch.clamp(psi, min=1.0e-14)
            b_squared = 2.0 * inverse_psi - 1.0 + torch.sqrt(2.0 * inverse_psi) * torch.sqrt(
                torch.clamp(2.0 * inverse_psi - 1.0, min=0.0)
            )
            quadratic = mean / (1.0 + b_squared) * (
                torch.sqrt(torch.clamp(b_squared, min=0.0)) + normal
            ).square()
            probability = (psi - 1.0) / (psi + 1.0)
            exponential = torch.where(
                uniform <= probability,
                torch.zeros_like(rate),
                mean / (1.0 - probability)
                * torch.log((1.0 - probability) / (1.0 - uniform)),
            )
            rate = torch.where(psi <= QE_PSI_CUTOFF, quadratic, exponential)
            integral += 0.5 * (previous + rate) * dt
            for row_index, exercise in schedule.get(step, ()):
                states[row_index, :, exercise] = rate[row_index]
                discounts[row_index, :, exercise] = torch.exp(-integral[row_index])

        immediate = torch.zeros_like(states)
        basis_state = torch.zeros_like(states)
        horizons = accrual[:, :, None] * (
            payment_axis[:, None, :] + 1 - exercise_axis[:, :, None]
        )
        bond_a, bond_b = cir_bond_coefficients(
            kappa[:, :, None], theta[:, :, None], sigma[:, :, None],
            torch.clamp(horizons, min=accrual[:, :, None]),
        )
        for exercise in range(MAX_EXERCISES):
            bonds = bond_a[:, exercise, :, None] * torch.exp(
                -bond_b[:, exercise, :, None] * states[:, None, :, exercise]
            )
            active = ((payment_axis >= exercise) & (payment_axis < payment_count[:, None]))[:, :, None]
            annuity = (active * accrual[:, :, None] * bonds).sum(dim=1)
            end_index = (payment_count - 1)[:, None, None].expand(-1, 1, num_paths)
            end_bond = torch.gather(bonds, 1, end_index).squeeze(1)
            immediate[:, :, exercise] = notional * torch.clamp(
                direction * (1.0 - end_bond - strike * annuity), min=0.0
            )
            basis_state[:, :, exercise] = (1.0 - end_bond) / annuity / BASIS_RATE_SCALE
        immediate *= (exercise_axis < exercise_count[:, None])[:, None, :]
        outputs.extend(price_summary(present_value_cashflows(
            immediate, basis_state, discounts, exercise_count
        )))
    synchronize(target)
    return outputs, {"wall_seconds": perf_counter() - started}


def price_batch(rows, model_by_id, product_by_id, *, num_paths, device,
                target_dt=1.0 / 52.0, batch_rows=None, **_):
    target = resolve_device(device)
    if batch_rows is None:
        batch_rows = (
            DEFAULT_CUDA_BATCH_ROWS
            if target.type == "cuda"
            else DEFAULT_CPU_BATCH_ROWS
        )
    if target.type == "cuda" and rows:
        warmup(target, shape=(min(batch_rows, len(rows)), num_paths), dtype=torch.float64)
        _price_batch_impl(
            rows[:batch_rows], model_by_id, product_by_id,
            num_paths=num_paths, target=target, target_dt=target_dt,
            batch_rows=batch_rows,
        )
    return _price_batch_impl(
        rows, model_by_id, product_by_id, num_paths=num_paths, target=target,
        target_dt=target_dt, batch_rows=batch_rows,
    )
